drivers: i2s: intel_s1000: Support for fractional divider to generate BCLK

sathishkuttan opened the pull request:
Fixes #7699

Commit:
arch: xtensa: intel_s1000: Reference clock API

Added a function to obtain the reference clock frequency value based on
SoC's bootstraps.
Added M/N divider base address in SoC header file

Signed-off-by: Sathish Kuttan <sathish.k.kuttan@intel.com>

Commit:
drivers: i2s: intel_s1000: Fractional bit-clock

Added support for fractional division of reference clock to obtain
I2S bit clocks
For example:
    If reference clock is 38.4MHz and
    if the I2S format is 2ch, 32bits per channel, 48KHz sample rate,
    the bit clock frequency has to be 2*32*48K = 3.072MHz
    38.4MHz/3.072MHz = 12.5 (a fraction)

Signed-off-by: Sathish Kuttan <sathish.k.kuttan@intel.com>

rgundi commented on drivers/i2s/i2s_cavs.c:
```diff
@@ -79,15 +79,16 @@ struct stream {
 	void *mem_block;
 	bool last_block;
-	int (*stream_start)(struct stream *, struct i2s_cavs_ssp *const,
+	int (*stream_start)(struct stream *, struct i2s_cavs_ssp *volatile,
```
What is the idea behind making this volatile instead of const? By const, i was ensuring the pointer variable doesn't get reassigned to anything else.

sathishkuttan replied:
Since the pointer is pointing to hardware registers, the idea is to force the compiler to read the register every time there is a read instead of re-using previously read values in the processor's machine registers.

rgundi replied:
If that's the case, it should be something like "volatile struct i2s_cavs_ssp *const". No? Here the struct is volatile where as the pointer to the struct remains a constant.

sathishkuttan replied:
That's correct. I'll update it.

rgundi reviewed (commented):
Some clarifications seeked.

sathishkuttan commented on drivers/i2s/i2s_cavs.c:
```diff
@@ -766,6 +771,7 @@ static void i2s1_irq_config(void)
 
 static const struct i2s_cavs_config i2s1_cavs_config = {
 	.regs = (struct i2s_cavs_ssp *)SSP_BASE(1),
+	.mn_regs = &((struct i2s_cavs_mn_div *)SSP_MN_DIV_BASE)[1],
```
It will translate into the address that is 8 bytes higher than `SSP_MN_DIV_BASE` because `sizeof(struct i2s_cavs_mn_div)` is 8 bytes.

Let me know if you want me to change this to be consistent with the `SSP_BASE(1)` statement one line above

rgundi replied:
Yes. Let's try to keep this uniform. Please change it to be like SSP_BASE(1).

sathishkuttan replied:
Changed.

Commit:
arch: soc: intel_s1000: set M/N divider ownership

Add bit definitions and set M/N divider ownership in
i2s_initialize.
Changes to comply with coding guidelines
Changes to address review comments

Signed-off-by: Sathish Kuttan <sathish.k.kuttan@intel.com>

rgundi commented:
Sathish, i still see many places where the old way of replacing const with volatile is still existing. I would rather prefer you modifying the existing commit instead of going in for a new commit in "change registers pointer type". That way it's easy for me to figure out what has changed from my code.

sathishkuttan commented:
@nashif Please merge when you get a chance.
